### DISCUSSION TITLE
Are all hash functions optimized using SIMD intrinsics?

### DISCUSSION THREAD

**Discussion by zxcvqwerasdf** [dotnet/runtime#109727] [Q&A] [answered]
For example SHA256/512

**Comment by teo-tsirpanis** [comment, marked as answer]
.NET does not implement cryptographic primitives by itself and instead relies on system libraries. Your question depends on the implementation of these libraries.

**Reply by zxcvqwerasdf** [reply]
> .NET does not implement cryptographic primitives by itself and instead relies on system libraries. Your question depends on the implementation of these libraries.

Thank for advice, checked in dnSpy, so it's BCrypt (SHA256.HashData)
```
Interop.BCrypt.NTSTATUS ntstatus = Interop.BCrypt.BCryptCreateHash(algHandle, out safeBCryptHashHandle, IntPtr.Zero, 0, key, key.Length, Interop.BCrypt.BCryptCreateHashFlags.None);
```

**Reply by teo-tsirpanis** [reply]
FYI the Windows implementation of the cryptographic algorithms lives in https://github.com/microsoft/SymCrypt and SHA256 for example [does use special instruction sets if available](https://github.com/microsoft/SymCrypt/blob/907622c6658877014ea3fea336e2efa407864fdc/lib/sha256.c#L1623-L1684). I don't know which Windows versions have which optimizations.
